ondevice_learning.qmd

```diff
@@ -1,61 +1,630 @@
+
+For example, a typical smartphone or Raspberry Pi may only have a few CPU cores, a few GB of RAM, and a small battery. These resources are fixed and can't easily be increased on demand like scaling cloud infrastructure. This reliance on edge devices directly restricts the complexity, efficiency, and size of models that can be deployed for on-device training:
+
+- **Complexity**: Limits on memory, computing, and power restrict model architecture design, constraining the number of layers and parameters.
+- **Efficiency**: Models must be heavily optimized through methods like quantization and pruning to run faster and consume less energy.
+- **Size**: Actual model files must be compressed as much as possible to fit within the storage limitations of edge devices.
+
+Thus, while the cloud offers endless scalability, on-device learning must operate within the tight resource constraints of endpoint hardware. This requires careful co-design of streamlined models, training methods, and optimizations tailored specifically for edge devices.
+
+#### Dataset Size
+
+In addition to limited computing resources, on-device learning is also constrained in terms of the dataset available for training models. 
+
+In the cloud, models are trained on massive, diverse datasets like ImageNet or Common Crawl. For example, ImageNet contains over 14 million images carefully categorized across thousands of classes. 
+
+On-device learning instead relies on smaller, decentralized data silos unique to each device. A smartphone camera roll may contain only thousands of photos centered around a user's specific interests and environments.
+
+
+This decentralized data leads to a lack of IID (independent and identically distributed) data. For instance, two friends may take many photos of the same places and objects, meaning their data distributions are highly correlated rather than independent.
+
+Reasons data may be non-IID in on-device settings:
+
+- **User heterogeneity**: different users have different interests and environments.
+- **Device differences:** sensors, regions, and demographics affect data.
+- **Temporal effects:** time of day, seasonal impacts on data.
+
+The effectiveness of machine learning relies heavily on large, diverse training data. With small, localized datasets, on-device models may fail to generalize across different user populations and environments.
+
+For example, a disease detection model trained only on images from a single hospital would not generalize well to other patient demographics. Without extensive, diverse medical images, the model's real-world performance would suffer.
+
+Thus, while cloud-based learning leverages massive datasets, on-device learning relies on much smaller, decentralized data silos unique to each user. This can limit model accuracy and generalization, especially for domain-specific use cases requiring broad training data.
+
+#### Accuracy & Generalization
```

Comment:
This content in this subsection appears to make very similar points to the dataset size section. Please either consolidate these or streamline them more so they are distinct.